OCaml_Intro_to_CS/part_6.ml

```diff
@@ -0,0 +1,220 @@
+(*
+Suppose you have two functions f : ‘a -> ‘b and g :
+‘c -> ‘a. Define their functional composition as follows:
+  *)
+
+let compose f g = fun x -> f (g x);;
+
+let inc x = x + 1;;
+
+let double x = x + x;;
+
+let h = compose inc double;;
+
+let result = h 5 ;;
+
+let () =
+  Printf.printf "result of compose inc double 5: %d\n" result;; 
+
+(*
+Please implement the compose_list function. Note: if the list is empty,
+then please return the argument (that is, be an identity function). In other words:
+# compose_list [] 7;;
+- : int = 7
+*)
+
+let dec x = x - 1;;
+
+let compose_list lst x =
+  let rec aux acc = function
+    | [] -> if acc = 0  then 
+      (
+        print_endline "Reached Identity";
+        x (* identity *)
+      )
+      else (
+        print_endline "Reached end of list without identity";
+        acc 
+      )
+    | h :: t -> 
+        if acc = 0 then 
+          (
+            print_endline "Applying function from list";
+            aux (h x) t 
+          )
+        else 
+          (
+            print_endline "Applying function from list with acc";
+            aux (h acc) t
+          )
+  in
+  aux 0 lst
+;;
+
+let operations = [inc ; double ; dec];;
+
+let res = compose_list operations 7;;
+
+let () = 
+  Printf.printf "Result of compose_list operations: %d\n" res;;
+
+(* for the extra 1 mark, using fold_left and fold_right *)
+let compose_list_fl func_lst x =
+  let f acc func = func acc in
+  let init = x in
+  List.fold_left f init func_lst
```

Comment:
Well this assignment sucks

Reply:
Sorry I think I gave you some fake news, the thing I provided is wrong

In terms of order of eval, it might be easier to visualize this:

```ocaml
Given [f1; f2; f3]

fold_left : f (f (f acc f1) f2) f3

fold_right : f f1 (f f2 (f f3 acc)))
```
So for the compose list, having 13 as the result means that you're evaluating from right to left (fold_right) to get the behaviour you want for `compose_lst_fl`.
```ocaml
let compost_lst_fl fn_l x = List.fold_left (fun acc f -> f acc) x l
```